docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,805 @@
+as an immediate failure, bypassing the timeouts and retries listed above, and
+the master attempts to remove the node from the cluster. Similarly, if a node
+detects that the elected master has disconnected then this is treated as an
+immediate failure, bypassing the timeouts and retries listed above, and the
+follower restarts its discovery phase to try and find or elect a new master.
+
+[float]
+[[modules-discovery-quorums]]
+=== Quorum-based decision making
+
+Electing a master node and changing the cluster state are the two fundamental
+tasks that master-eligible nodes must work together to perform. It is important
+that these activities work robustly even if some nodes have failed, and
+Elasticsearch achieves this robustness by only considering each action to have
+succeeded on receipt of responses from a _quorum_, a subset of the
+master-eligible nodes in the cluster. The advantage of requiring only a subset
+of the nodes to respond is that it allows for some of the nodes to fail without
+preventing the cluster from making progress, and the quorums are carefully
+chosen so as not to allow the cluster to "split brain", i.e. to be partitioned
+into two pieces each of which may make decisions that are inconsistent with
+those of the other piece.
+
+Elasticsearch allows you to add and remove master-eligible nodes to a running
+cluster. In many cases you can do this simply by starting or stopping the nodes
+as required, as described in more detail in the
+<<modules-discovery-adding-removing-nodes,section on adding and removing
+nodes>>.
+
+As nodes are added or removed Elasticsearch maintains an optimal level of fault
+tolerance by updating the cluster's _voting configuration_, which is the set of
+master-eligible nodes whose responses are counted when making decisions such as
+electing a new master or committing a new cluster state. A decision is only made
+once more than half of the nodes in the voting configuration have responded.
+Usually the voting configuration is the same as the set of all the
+master-eligible nodes that are currently in the cluster, but there are some
+situations in which they may be different.
+
+To be sure that the cluster remains available you **must not stop half or more
+of the nodes in the voting configuration at the same time**. As long as more
+than half of the voting nodes are available the cluster can still work normally.
+This means that if there are three or four master-eligible nodes then the
+cluster can tolerate one of them being unavailable; if there are two or fewer
+master-eligible nodes then they must all remain available.
+
+After a node has joined or left the cluster the elected master must issue a
+cluster-state update that adjusts the voting configuration to match, and this
+can take a short time to complete. It is important to wait for this adjustment
+to complete before removing more nodes from the cluster.
 
 [float]
-=== Settings
+==== Setting the initial quorum
```

Comment:
I think this conflicts a bit with the boostrap section - there is some duplication and it is related. Should these be merged?

Reply:
I added a link and reorganised a bit in 1d69b0a268d585c5fd280157458de44fbabd91bc. However I do think these should be separate; this section describes much lower-level details, in terms of quorums, which means it depends on understanding `Quorum-based decision making`; the section on bootstrapping is supposed to be higher-level and doesn't require such a deep understanding to take useful action.